### ISSUE TITLE
[ceremonies] improve testability

### ISSUE THREAD

**Issue by brenzi** [encointer/pallets#171]
the following functions should be stateless for better unit testing:
* no access to pallet storage
* not using any functions that have access to pallet storage
* => all information must be passed as arguments

https://github.com/encointer/pallets/blob/05b4ff16e8b956c1d460165c3583008b9edb0a1b/ceremonies/src/lib.rs#L1342

https://github.com/encointer/pallets/blob/05b4ff16e8b956c1d460165c3583008b9edb0a1b/ceremonies/src/lib.rs#L1243

https://github.com/encointer/pallets/blob/05b4ff16e8b956c1d460165c3583008b9edb0a1b/ceremonies/src/lib.rs#L985

... and more

**Comment by pifragile** [comment]
I propose the following algorithm to validate a meetup:

1. build hashmap AccountId -> {vote:u32 , attestations: Vec\<AccountId>}
2. compute majority vote result from hashmap (iterate over all keys, gather the votes, compute result)
3. exclude all participants that do not vote like majority
4. exclude all participants that have given too few attestations(such that they dont receive reward, but also no attestations from them are expected for other participants)
5. check attestations for the remaining participants, using a given threshold on number of attestations necessary
6. pay rewards


Remarks:

-Steps 2-5 happen in the new extranal crate

-I think that it makes more sense to keep the votes in the hashmap, because of step 3, like this we only need to fetch the data from storage once (@brenzi if you would like to separate the datastructures, we can also have two hashmaps, one for attestations and one for votes)

-Step 4. fixes the DOS attack described in #214 
-The thresold described in 5. can fix #168 

-It is more intuitive to use AccoundId than ParticipantIndex, because the attestation storage uses AccountId. Do we really want our external crate to only work with indices? We can do it, but it requires another conversion step.

@brenzi @clangenb please comment

PS: The point i wanted to make yesterday in the call (that we should take as vote simply the number of sent attestations) was that it would render step 4 obsolete, because accounts that dont send enough attestations would already be excluded by the majority vote. But if we want to introduce  some tolerance there as well (because we dont want to exclude somebody who couldnt attest another person, because they have a broken screen for example), my argument is invalid.

**Comment by pifragile** [comment]
> I just wonder; if we are that strict in 4, we can simply check this when the claims are submitted, and disallow registering claims with `vote != attestationCount` in the first place.

yes this is exactly the dilemma described in the PS. I think we want to allow some tolerance also there in order for the scenario with somebody having a broken screen.

so we would have to define a threshold T_attack below somebody is considered an attacker in 4.
in 5 we would need two thresolds: minimum number of incoming attestations(T_out) and minimum number of outgoing attestations(T_in).

Example:

Meetup with 10 people

All of them vote that 10 people were present
The screen of A is a little broken and B cannot scan it, but the others can.
C is an attacker and does not send any attestations

So now A has 8 incoming attestations and 10 outgoing
B has 9 incoming attestations and 9 outgoing
C has 10 incoming attestations and 0 outgoing
all other participants have 9 incoming and 10 outgoing

when we now set T_attack to 1, T_incoming to 8 and T_outgoing to 9, everybody but the attacker gets their reward.

We can experiment with those parameters.
1 simplification that comes to mind would be to set T_attack = T_outgoing.

**Comment by brenzi** [comment]
Ballotting votes doesn't need to know attestations. So I suggest to split the data structure into votea and attestations. That makes testing ballotting easier

**Comment by clangenb** [comment]
This is implemented, no?
